ipn/ipnlocal/state_test.go

```diff
@@ -0,0 +1,778 @@
+	cc.called("SetHostinfo")
+}
+
+func (cc *mockControl) SetNetInfo(ni *tailcfg.NetInfo) {
+	cc.called("SetNetinfo")
+	cc.logf("SetNetInfo: %v", *ni)
+	cc.called("SetNetInfo")
+}
+
+func (cc *mockControl) UpdateEndpoints(localPort uint16, endpoints []tailcfg.Endpoint) {
+	// validate endpoint information here?
+	cc.logf("UpdateEndpoints: lp=%v ep=%v", localPort, endpoints)
+	cc.called("UpdateEndpoints")
+}
+
+// A very precise test of the sequence of function calls generated by
+// ipnlocal.Local into its controlclient instance, and the events it
+// produces upstream into the UI.
+//
+// [apenwarr] Normally I'm not a fan of "mock" style tests, but the precise
+// sequence of this state machine is so important for writing our multiple
+// frontends, that it's worth validating it all in one place.
+//
+// Any changes that affect this test will most likely require carefully
+// re-testing all our GUIs (and the CLI) to make sure we didn't break
+// anything.
+//
+// Note also that this test doesn't have any timers, goroutines, or duplicate
+// detection. It expects messages to be produced in exactly the right order,
+// with no duplicates, without doing network activity (other than through
+// controlclient, which we fake, so there's no network activity there either).
+//
+// TODO: A few messages that depend on magicsock (which actually might have
+// network delays) are just ignored for now, which makes the test
+// predictable, but maybe a bit less thorough. This is more of an overall
+// state machine test than a test of the wgengine+magicsock integration.
+func TestStateMachine(t *testing.T) {
+	assert := assert.New(t)
+
+	logf := t.Logf
+	store := new(ipn.MemoryStore)
+	e, err := wgengine.NewFakeUserspaceEngine(logf, 0)
+	if err != nil {
+		t.Fatalf("NewFakeUserspaceEngine: %v", err)
+	}
+
+	cc := newMockControl()
+	ccGen := func(opts controlclient.Options) (controlclient.Client, error) {
+		cc.mu.Lock()
+		cc.opts = opts
+		cc.logf = opts.Logf
+		cc.authBlocked = true
+		cc.mu.Unlock()
+
+		cc.called("New")
+		return cc, nil
+	}
+	b, err := NewLocalBackendWithClientGen(logf, "logid", store, e, ccGen)
+	if err != nil {
+		t.Fatalf("NewLocalBackend: %v", err)
+	}
+
+	notifies := notifyThrottler{t: t}
+	notifies.expect(0)
+
+	b.SetNotifyCallback(func(n ipn.Notify) {
+		if n.State != nil ||
+			n.Prefs != nil ||
+			n.BrowseToURL != nil ||
+			n.LoginFinished != nil {
+			logf("\n%v\n\n", n)
+			notifies.put(n)
+		} else {
+			logf("\n(ignored) %v\n\n", n)
+		}
+	})
+
+	// Check that it hasn't called us right away.
+	// The state machine should be idle until we call Start().
+	assert.Equal(cc.getCalls(), []string{})
+
+	// Start the state machine.
+	// Since !WantRunning by default, it'll create a controlclient,
+	// but not ask it to do anything yet.
+	notifies.expect(2)
+	assert.Nil(b.Start(ipn.Options{
+		StateKey: ipn.GlobalDaemonStateKey,
+	}))
+	{
+		// BUG: strictly, it should pause, not unpause, here, since !WantRunning.
```

Comment:
Why these indents? Feels like it's visually trying to tell me that this is one test case, but then there's also outdented code sprinkled between them, which makes my brain skip a beat. Do I have the intent right?

Reply:
Well, one reason is so that the "outputs" part of one section doesn't accidentally leak into the conditionals of the next section. In a function this long, all the local variables start to add up, and this keeps it simpler.

I would kind of like to take each of the sections and convert them to functions, then call the functions in different sequences, rather than cut-and-paste-and-slightly-modify like I did here. But it's not as simple as it sounds, because of the bugs and inconsistencies in the code under test. (ie. you might think that every time you call Logout, you could check that WantRunning=false. But it turns out that is currently only sometimes true.) I think it'll get easier once the actual bugs are removed.